bayes_opt1/util.py

```diff
@@ -1,10 +1,233 @@
+
+    if type(x) is not np.ndarray:
+        sys.exit('Error in _kg4: if p>1, x needs to be a NP.ndarray')
+    if len(x.shape) > 1 and x.shape[0] > 1:
+        sys.exit('Error in _kg4: x needs to have 1 row and p columns.')
+    
+    
+    
+    def mu_x_star_hat(x_obs,y_obs,gp,x_star_hat,mean,std,i):
+        '''
+        see Algorithm 4:
+        In our setting, we apply this by letting x_star_hat be the x^prime maximizing µ_{n+1}(x^prime; x, µ_n(x)+σ_n(x)Z), 
+        and then calculating the gradient of µ_{n+1}(x_star_hat; x, µ_n(x)+σ_n(x)Z) with respect to x 
+        while holding x_star_hat fixed.
+        '''
+        y=np.random.normal(loc=mean, scale=std)
+        x_newobs=np.concatenate((x_obs, i.reshape(1,-1)), axis = 0)
+        y_newobs=np.append(y_obs,y)
+        mu=posterior(gp, x_newobs, y_newobs, x_star_hat)
+        return np.asscalar(mu)
+    
+    dim = optimizer._space.bounds.shape[0] # size of state space (e.g. if we are working on R^3, n = 3) 
+    x_obs_temp = np.array([[res["params"]] for res in optimizer.res]) 
+    y_obs = np.array([res["target"] for res in optimizer.res])
+    # x_obs needs to be np.array of size (n,p)
+    # y_obs needs to be np.array of size (n,)
+    # see https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html
+    x_obs = x_obs_to_array(x_obs_temp, dim)
+
+    grid = grid_construction(optimizer._space.bounds, n_grid)
+    
+    temp = find_mean_max(x_obs,y_obs,gp,grid)
+    mean_nstar = temp[0]
+
+    grad_temp = []
+    mean, std= gp.predict(x.reshape(1,-1), return_std=True)
+    for j in range(J):
+        print("j = ",j+1)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            temp = find_mean_max(x_obs,y_obs,gp,grid,mean,std,x)
+            mean_n1star = temp[0]
+            x_star_hat = grid[temp[1]].reshape(1,-1)
+            def func(w):
+                return mu_x_star_hat(x_obs,y_obs,gp,x_star_hat,mean,std,w)
+            
+            grad_temp.append(nd.Gradient(func)(x))
+    grad = np.zeros(dim)
+    for i in grad_temp:
+        grad = grad + i
+    grad = grad/J
+    return grad
+
+
+
+# Implementation algorithm 3 (Frazier): Knowledge gradient 
+def minimize_kg_sgd(R, T, a, pbounds, optimizer, gp):
+
+    dim = pbounds.shape[0]
+    x_T = []
+    
+    for r in range(R):
+        x_t = our_lhs(1, pbounds)[0]
+        for t in range(T):
+            G = _kg4(x_t, optimizer, gp, n_grid = 10, J = 2) 
+            alpha = a/(a+t+1)
+            x_t_new = x_t + alpha * G
+            #print("G =",G)
+            #print("x0 =",x0)
+            #print("x_new =", x_new)
+            x_t = x_t_new
+        x_T.append(x_t_new)
+
+    kg = np.zeros(len(x_T))
+    
+    for i in range(len(x_T)):
+        #print("x_T =", x_T[i])
+        kg[i] = _kg2(x_T[i].reshape(1,-1), optimizer, gp, n_grid = 10, J = 2)
+        #print("kg =", kg[i])
+
+    return x_T[np.argmax(kg)] # WITHOUT reshape
```

Comment:
Questa è la nuova funzione che gestisce sia il caso unidimensionale che più dimensioni. Abbiamo testato nel file "TEST" e va, abbiamo però per ora usato n_grid e gli altri parametri molto piccoli per avere i risultati in tempi ragionevoli.
Abbiamo lasciato i print che abbiamo usato se vuole fare qualche prova... altrimenti li togliamo!

Reply:
Mi piace molto questa nuova versione, è molto più pulita! Togliete pure tutti i print, ho visto e sembra funzionare tutto